### ISSUE TITLE
[java] UnnecessaryLocalBeforeReturn false positive with variable captured by method reference

### ISSUE THREAD

**Issue by oowekyala** [pmd/pmd#2167]
For example:

```java
    public Supplier<Object> create() {
        final Object o = new Object();
        return o::clone; // violation
    }
```

`o` is captured, and keeping its declaration separate keeps the code readable (more so with a more complicated RHS).

**Comment by Fernal73** [comment]
Your code doesn't compile since clone has protected access. You might want to do new Object[] instead.

**Comment by adangel** [comment]
Indeed, the original example doesn't compile, but we can just use `toString()` instead, as in the test case in #2168:

```java
public class ObjectCreator {
    public Supplier<String> create() {
        final Object o = new Object(); // captured by the method ref
        return o::toString;
    }
```

**Comment by Fernal73** [comment]
It doesn't really matter. PMD doesn't care...
Unless you wish to institute a policy that code examples provided must
compile....




-- 
Thanks,
Linus Fernandes.
